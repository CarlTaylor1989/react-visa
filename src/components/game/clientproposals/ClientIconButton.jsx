import React from 'react';
import PropTypes from 'prop-types';
import { I18n, Translate } from 'react-redux-i18n';
import ClientLocked from './ClientLocked';
import ClientIcon from './ClientIcon';

import './ClientIconButton.scss';

const ClientIconButton = ({
  client,
  disabled,
  locked,
  onClientSelected,
  selected,
  completed
}) => {
  const problemType = I18n.t(`clientprops.clients.${client}.problemType`);
  let extraClasses = selected ? ' selected' : '';
  extraClasses = completed ? ' completed' : extraClasses;
  extraClasses = locked && locked.id ? ' locked' : extraClasses;

  return (
    <button
      type="button"
      className={`clientIconButton${extraClasses}`}
      onClick={() => onClientSelected(client)}
      disabled={disabled || (locked && locked.id)}
    >
      <ClientIcon client={client} />
      {locked && locked.id && (
        <ClientLocked client={client} locked={locked} />
      )}
      <div className="infoBlock">
        <div className="nameType">
          <Translate className="clientName" value={`clientprops.clients.${client}.name`} />
          <Translate className="clientType" value={`clientprops.problems.${problemType}.cpBtn`} />
        </div>
      </div>
    </button>
  );
};

ClientIconButton.defaultProps = {
  locked: {}
};

ClientIconButton.propTypes = {
  client: PropTypes.string.isRequired,
  disabled: PropTypes.bool.isRequired,
  locked: PropTypes.object,
  onClientSelected: PropTypes.func.isRequired,
  selected: PropTypes.bool.isRequired,
  completed: PropTypes.bool.isRequired
};

export default ClientIconButton;
