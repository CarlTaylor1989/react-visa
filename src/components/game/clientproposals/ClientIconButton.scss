@import '../../../styles/buttons';
@import '../../../styles/clients';

$client-icon-height: 130px;
$client-icon-width: 290px;
$client-icon-padding: 10px;
$client-icon-image-height: 126px;
$tick-icon: '../../../styles/icons/clients/tick_icon.svg';
$tick-icon-width: 26px;
$tick-icon-height: 20px;
$tick-icon-diameter: 46px;

.clientIconButton {
  @extend %button-reset;
  @include align-items(center);
  box-sizing: content-box;
  color: $colour-white;
  cursor: pointer;
  display: flex;
  flex-direction: row;
  font-size: 16px;
  height: $client-icon-height;
  margin-bottom: 10px;
  margin-right: 5px;
  padding: $client-icon-padding;
  position: relative;
  width: $client-icon-width;

  .infoBlock {
    display: flex;
    flex: 1;
    flex-direction: column;
    margin-left: 20px;

    .clientName,
    .clientType {
      display: block;
      font-size: 20px;
      text-align: center;
    }

    .clientName {
      font-weight: 600;
    }
  }

  &::before {
    @include border-radius(20px);
    @include multi-angle-gradient(
      152deg,
      rgba($colour-white, .05) 0%,
      rgba($colour-white, .1) 38.49%,
      rgba($colour-white, .18) 79.27%,
      rgba($colour-white, .21) 100%
    );
    @include opacity(0);
    border: 2px solid $colour-white;
    bottom: 0;
    content: '';
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transition: opacity .5s ease-in-out;
    z-index: 1;
  }

  &.selected,
  &:hover {
    &::before {
      @include opacity(1);
    }
  }

  &:disabled {
    &::before {
      display: none;
    }
  }

  &.completed {
    .infoBlock {
      .productIndicator {
        .productIndicatorCircle {
          @include box-shadow(0, 0, 20px, 4px, $client-progress);
        }
      }
    }

    .clientIcon {
      @include border-radius($client-icon-image-height / 2);
      border: 4px solid $colour-yellow;
      position: relative;

      &::before {
        // pseudo element for gradient behind main icon
        @include border-radius($client-icon-image-height / 2);
        background: $colour-button;
        content: '';
        height: calc(100% - 10px);
        left: 5px;
        position: absolute;
        top: 5px;
        width: calc(100% - 10px);
        z-index: -1;
      }

      &::after {
        // pseudo element for gradient behind tick icon (background image)
        @include border-radius($tick-icon-diameter / 2);
        background-color: $colour-button;
        background-image: url($tick-icon);
        background-position: center;
        background-repeat: no-repeat;
        border: 2px solid $colour-white;
        box-sizing: border-box;
        content: '';
        height: $tick-icon-diameter;
        left: auto;
        position: absolute;
        right: 0;
        top: 0;
        width: $tick-icon-diameter;
        z-index: 1;
      }
    }
  }

  &.locked {
    cursor: not-allowed;

    .clientIcon,
    .infoBlock {
      @include opacity(.3);
    }
  }

  .clientLocked {
    left: 9px;
    position: absolute;
    top: 9px;
  }
}
