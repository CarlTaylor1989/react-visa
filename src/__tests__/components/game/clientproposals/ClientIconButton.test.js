/* global toJson */
import React from 'react'; // eslint-disable-line no-unused-vars
import ClientIconButton from '../../../../components/game/clientproposals/ClientIconButton';

let mockProps;
beforeEach(() => {
  mockProps = {
    client: 'lv1cl1',
    completed: false,
    disabled: false,
    locked: {},
    reqAttempts: 1,
    selected: false,
    onClientSelected: jest.fn()
  };
});

test('renders the ClientIconButton component with the selected state', () => {
  const wrapper = shallow(<ClientIconButton {...mockProps} selected />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('renders the ClientIconButton component with the disabled state', () => {
  const wrapper = shallow(<ClientIconButton {...mockProps} disabled />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('renders the ClientIconButton component with the completed state', () => {
  const wrapper = shallow(<ClientIconButton {...mockProps} completed />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('renders the ClientIconButton component with the locked state', () => {
  const wrapper = shallow(<ClientIconButton {...mockProps} locked={{ id: 'irrelevant' }} />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('button calls onClientSelected() method when selected', () => {
  const wrapper = shallow(<ClientIconButton {...mockProps} />);
  wrapper.find('.clientIconButton').at(0).simulate('click');
  expect(mockProps.onClientSelected).toHaveBeenCalled();
});
