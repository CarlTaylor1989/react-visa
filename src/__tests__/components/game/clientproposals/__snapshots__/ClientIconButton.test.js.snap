// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the ClientIconButton component with the completed state 1`] = `
<button
  className="clientIconButton completed"
  onClick={[Function]}
  type="button"
>
  <ClientIcon
    client="lv1cl1"
    proposal={false}
  />
  <div
    className="infoBlock"
  >
    <div
      className="nameType"
    >
      <Translate
        className="clientName"
        tag="span"
        value="clientprops.clients.lv1cl1.name"
      />
      <Translate
        className="clientType"
        tag="span"
        value="clientprops.problems.problemType.cpBtn"
      />
    </div>
  </div>
</button>
`;

exports[`renders the ClientIconButton component with the disabled state 1`] = `
<button
  className="clientIconButton"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  <ClientIcon
    client="lv1cl1"
    proposal={false}
  />
  <div
    className="infoBlock"
  >
    <div
      className="nameType"
    >
      <Translate
        className="clientName"
        tag="span"
        value="clientprops.clients.lv1cl1.name"
      />
      <Translate
        className="clientType"
        tag="span"
        value="clientprops.problems.problemType.cpBtn"
      />
    </div>
  </div>
</button>
`;

exports[`renders the ClientIconButton component with the locked state 1`] = `
<button
  className="clientIconButton locked"
  disabled="irrelevant"
  onClick={[Function]}
  type="button"
>
  <ClientIcon
    client="lv1cl1"
    proposal={false}
  />
  <Connect(ClientLocked)
    client="lv1cl1"
    locked={
      Object {
        "id": "irrelevant",
      }
    }
  />
  <div
    className="infoBlock"
  >
    <div
      className="nameType"
    >
      <Translate
        className="clientName"
        tag="span"
        value="clientprops.clients.lv1cl1.name"
      />
      <Translate
        className="clientType"
        tag="span"
        value="clientprops.problems.problemType.cpBtn"
      />
    </div>
  </div>
</button>
`;

exports[`renders the ClientIconButton component with the selected state 1`] = `
<button
  className="clientIconButton selected"
  onClick={[Function]}
  type="button"
>
  <ClientIcon
    client="lv1cl1"
    proposal={false}
  />
  <div
    className="infoBlock"
  >
    <div
      className="nameType"
    >
      <Translate
        className="clientName"
        tag="span"
        value="clientprops.clients.lv1cl1.name"
      />
      <Translate
        className="clientType"
        tag="span"
        value="clientprops.problems.problemType.cpBtn"
      />
    </div>
  </div>
</button>
`;
